README.md
Jira Cookbook
==================
[![Chef cookbook](https://img.shields.io/cookbook/v/jira.svg)](https://supermarket.chef.io/cookbooks/jira)
[![Build Status](https://secure.travis-ci.org/afklm/jira.png?branch=master)](http://travis-ci.org/afklm/jira)

*WARNING* - This cookbook was essentially replaced by a new cookbook starting version 2.0. This is a breaking change.

## Description

Installs/Configures Atlassian JIRA. Please see [COMPATIBILITY.md](COMPATIBILITY.md) for more information about JIRA releases that are tested and supported by this cookbook and its versions.

Starting from release 2.8.0 of this cookbook, the default DB used is Postgres due to various bugs and speed issues with MySQL.

## Requirements

### Chef

* Chef 11.14+ for version 2.7.1+ of this cookbook

### Platforms

* CentOS 6
* RHEL 6
* Ubuntu 12.04

### Databases

* MySQL
* Postgres

### JDK/JRE

The Atlassian JIRA Linux installer will automatically configure a bundled JRE. If you wish to use your own JDK/JRE, with say the `java` cookbook, then as of this writing it must be Oracle and version 1.7 or 1.8.

For the up-to-date list, please see [Supported Platforms](https://confluence.atlassian.com/display/JIRA/Supported+Platforms)

Necessary configuration with `java` cookbook:
* `node['java']['install_flavor'] = "oracle"`
* `node['java']['oracle']['accept_oracle_download_terms'] = true`
* `recipe[java]`

A /ht to [@seekely](https://github.com/seekely) for the [documentation nudge](https://github.com/bflad/chef-jira/issues/2).

## Attributes

These attributes are under the `node['jira']` namespace.

Attribute    | Description                                           | Type   | Default
-------------|-------------------------------------------------------|--------|---------------------------------------
checksum     | SHA256 checksum for JIRA install                      | String | auto-detected by helper method
home_path    | home directory for JIRA                               | String | /var/atlassian/application-data/jira
install_path | location to install JIRA                              | String | /opt/atlassian/jira
install_type | JIRA install type - "installer" or "standalone"       | String | installer
init_type    | JIRA init service type - "sysv"                       | String | sysv
url          | URL for JIRA install                                  | String | auto-detected by helper method
user         | user running JIRA                                     | String | jira
version      | JIRA version to install                               | String | 7.0.4
flavor       | JIRA product flavor to install - 'core' or 'software' | String | software

**Notice:** If `['jira']['install_type']` is set to `installer`, then the installer will try to upgrade your JIRA instance located in `['jira']['install_path']` (if it exists) to the `['jira']['version']`.

If you want to avoid an unexpected upgrade, just set or override `['jira']['version']` attribute value to that of your current JIRA version.

### JIRA Database Attributes

All of these `node['jira']['database']` attributes are overridden by `jira/jira` encrypted data bag (Hosted Chef) or data bag (Chef Solo), if it exists

Attribute | Description                                   | Type   | Default
----------|-----------------------------------------------|--------|------------
host      | FQDN or "127.0.0.1"                           | String | 127.0.0.1
name      | JIRA database name                            | String | jira
password  | JIRA database user password                   | String | changeit
type      | JIRA database type - "mysql", or "postgresql" | String | postgresql
user      | JIRA database user                            | String | jira

The Postgres DB is automatically tuned for 'web' and with 1GB of memory for Postgres. Please see the https://github.com/hw-cookbooks/postgresql cookbook for more details, specifically the config_pgtune section.

Please note that specifying "127.0.0.1" for the host automatically installs `['database']['type']` server in the default recipe.

### JIRA JVM Attributes

These attributes are under the `node['jira']['jvm']` namespace.

Attribute       | Description                                                                       | Type   | Default
----------------|-----------------------------------------------------------------------------------|--------|--------
minimum_memory  | JVM minimum memory                                                                | String | 512m
maximum_memory  | JVM maximum memory                                                                | String | 768m
maximum_permgen | JVM maximum PermGen memory                                                        | String | 256m
java_opts       | additional JAVA_OPTS to be passed to JIRA JVM during startup                      | String | ""
support_args    | additional JAVA_OPTS recommended by Atlassian support for JIRA JVM during startup | String | ""

### JIRA Tomcat Attributes

These attributes are under the `node['jira']['tomcat']` namespace.

Attribute | Description      | Type   | Default
----------|------------------|--------|--------
port      | Tomcat HTTP port | Fixnum | 8080

## Recipes

* `recipe['jira']` 'Installs/configures Atlassian JIRA'
* `recipe['jira::apache2']` 'Installs/configures Apache 2 as proxy (ports 80/443)'
* `recipe['jira::container_server_configuration']` 'Configures container server for JIRA deployment'
* `recipe['jira::database']` 'Installs/configures MySQL/Postgres server, database, and user for JIRA'
* `recipe['jira::installer']` 'Installs/configures JIRA via installer'
* `recipe['jira::standalone']` 'Installs/configures JIRA via standalone archive'

## Usage

### JIRA Server Data Bag

Optionally for securely overriding attributes on Hosted Chef, create a `jira/jira` encrypted data bag with the model below. Chef Solo can override the same attributes with a `jira/jira` unencrypted data bag of the same information.

_required:_
* `['database']['type']` - "mssql", "mysql", or "postgresql"
* `['database']['host']` - FQDN or "localhost" (localhost automatically installs `['database']['type']` server)
* `['database']['name']` - Name of JIRA database
* `['database']['user']` - JIRA database username
* `['database']['password']` - JIRA database username password

_optional:_
* `['database']['port']` - Database port, defaults to standard database port for `['database']['type']`

Repeat for other Chef environments as necessary. Example:

    {
      "id": "jira",
      "development": {
        "database": {
          "type": "postgresql",
          "host": "localhost",
          "name": "jira",
          "user": "jira",
          "password": "jira_db_password",
        }
      }
    }

### Default JIRA Installation

The simplest method is via the default recipe, which uses `node['jira']['install_type']` (defaults to installer).

* Optionally (un)encrypted data bag or set attributes
  * `knife data bag create jira`
  * `knife data bag edit jira jira --secret-file=path/to/secret`
* Add `recipe[jira]` to your node's run list.

### Standalone JIRA Installation

Operates similarly to installer installation, however has added benefits of using `ark` to create version symlinks of each install. Easily can rollback upgrades by changing `node['jira']['version']`.

* Optionally (un)encrypted data bag or set attributes
  * `knife data bag create jira`
  * `knife data bag edit jira jira --secret-file=path/to/secret`
* Set `node['jira']['install_type']` to standalone
* Add `recipe[jira]` to your node's run list.

### Custom JIRA Configurations

Using individual recipes, you can use this cookbook to configure JIRA to fit your environment.

* Optionally (un)encrypted data bag or set attributes
  * `knife data bag create jira`
  * `knife data bag edit jira jira --secret-file=path/to/secret`
* Add individual recipes to your node's run list.

## Testing and Development

* Quickly testing with Vagrant: [VAGRANT.md](VAGRANT.md)
* Full development and testing workflow with Test Kitchen and friends: [TESTING.md](TESTING.md)

For Vagrant, you may need to add the following hosts entries:

* 192.168.50.10 jira-centos-6
* 192.168.50.10 jira-ubuntu-1204
* (etc.)

The running JIRA server is then accessible from the host machine:

CentOS 6 Box:
* Web UI (installer/standalone): https://jira-centos-6/

Ubuntu 12.04 Box:
* Web UI (installer/standalone): https://jira-ubuntu-1204/

## Contributing

Please see contributing information in: [CONTRIBUTING.md](CONTRIBUTING.md)

## Maintainers

* KLM Royal Dutch Airlines

## Sponsors

Development of this cookbook has been generously supported in part by
the code contributions of the following organizations:

* [University of Pennsylvania](http://www.upenn.edu/)
  ([@bflad](https://github.com/afklm/jira/commits?author=bflad))
* [KLM Royal Dutch Airlines](https://www.klm.com/)
  ([@mvdkleijn](https://github.com/afklm/jira/commits?author=mvdkleijn))
* [Parallels Inc.](https://www.parallels.com/)
  ([@legal90](https://github.com/afklm/jira/commits?author=legal90))
* [Blended Perspectives Inc.](http://www.blendedperspectives.com/)
  ([@patcon](https://github.com/afklm/jira/commits?author=patcon))

## License

Please see licensing information in: [LICENSE](LICENSE)


CHANGELOG.md
## x.y.z (pending)

## 2.7.3

* Added JIRA 7.0.2
* Bumped default version to 7.0.2

## 2.7.2

* Added sensitive true for crowd_sso template
* Set Postgres DB owner to JIRA user
* Explicitly set `home_path` perms.
  [[#48]](https://github.com/afklm/jira/issues/48)
* Converted `virtual_host_name` and `virtual_host_alias` derived attrs
  into helper methods for templates, etc.
[[#45]](https://github.com/afklm/jira/issues/45)

## 2.7.0

* Added support for JIRA 7.0
* Added `node['jira']['flavor']` attribute with default value of 'software'
* Bump default JIRA version to 7.0.0

## 2.6.3

* Add JIRA 6.4.12 and bump default version

## 2.6.2

* Added a `node['jira']['group']` attribute for clarity and override ability
* Issue warn in library for unsupported DB types to make override possible

## 2.6.1

* Removed unused config entries from cookbook metadata
* Replaced file cookbook with FileEdit#search_file_replace

## 2.6.0

* Redirect http-based requests to https
* Auto upgrade if `node['jira']['version']` is higher than installed version
* Configures Crowd SSO if `node['jira']['crowd_sso']['enabled']` is true
* Improve file restrictions for standalone installs

## 2.5.1

* MIGRATED: renamed cookbook chef_jira -> jira after getting the supermarket
            namespace
* Bump default version to JIRA 6.4.11
* Use https for Jira downloads.
  [[GH-18]](https://github.com/afklm/jira/issues/18)
* Added service restart when ark resource changes.
  [[GH-16]](https://github.com/afklm/jira/issues/16)
* Added support for PostgreSQL 9.2+ in `dbconfig.xml`.
  [[GH-14]](https://github.com/afklm/jira/issues/14)
* Fixed LWRPs after cookbook name change.
  [[GH-13]](https://github.com/afklm/jira/pull/13)
* Set Tomcat `proxyName`/`proxyPort` even without SSL.
  [[GH-11]](https://github.com/afklm/jira/issues/11)
* Removed unnecessary non-dynamic `web.xml` template.
  [[GH-10]](https://github.com/afklm/jira/issues/10)
* Fixed setting of `jira.home` for all install types.
  [[GH-15]](https://github.com/afklm/jira/issues/15)

Thanks go to @elijah @gsreynolds @legal90 and @patcon for helping out in this
release.

## 2.1.0

* MIGRATED: renamed cookbook jira -> chef_jira

## 2.0.1

* Bugfix: #8: Remove include_attribute from default attributes to prevent hard dependency on tomcat cookbook

## 2.0.0

If you were using the default recipe, there are no changes you need to make for your environment to upgrade this cookbook.

Major features are full support for standalone deployments and war building/deployments. Using ark where possible (its worked well for those using my Stash cookbook).

I've removed the upgrade recipe in favor of using ark. If you'd like to keep old JIRA installations around (in case of upgrade issues, etc.), I would recommend switching to the standalone install_type since ark will automatically create versioned symlinks and you can easily revert `node['jira']['version']`. To convert install_type if install_path is the default (X.Y.Z being currently installed `node['jira']['version']`):
* `service jira stop`
* `mv /opt/atlassian/jira /opt/atlassian/jira-X.Y.Z`
* `ln -s /opt/atlassian/jira-X.Y.Z /opt/atlassian/jira`
* Set `node['jira']['install_type']` to standalone
* Run Chef Client

Other than that, migrated some recipes/templates and split out some recipes from the old linux_installer recipe, so ensure that if you're using a custom run list or template override for any nodes, that they include the new recipes/template location as necessary.

Full details:
* REMOVED: upgrade recipe and associated backup_* and *_backup attributes
* MIGRATED: linux_installer -> installer recipe
* MIGRATED: tomcat_configuration -> container_server_configuration recipe
* MIGRATED: Tomcat templates into tomcat folder:
  * permgen.sh.erb -> tomcat/permgen.sh.erb
  * server.xml.erb -> tomcat/server.xml.erb
  * setenv.sh.erb -> tomcat/setenv.sh.erb
  * web.xml.erb -> tomcat/web.xml.erb
* SPLIT: SysV init service configuration into sysv recipe and add init_type attribute
* SPLIT: database jar deployment (mysql_connector_j, etc.) into container_server_jars recipe which also installs JIRA jars for war install_type
* Bugfix: Use :create action instead of :create_if_missing for installer remote_file
* Enhancement: Add standalone and war recipes and quite a few attributes for supporting those install_type's
* Enhancement: Add build_war recipe
* Enhancement: LWRPs for handling multiple instances of install, etc.
* Enhancement: Bump default JIRA version to 6.1.5

## 1.7.0

* Bump default JIRA version to 6.1

## 1.6.0

* Bump default JIRA version to 6.0.7

## 1.5.0

* Bump default JIRA version to 6.0.6

## 1.4.0

* Initial Microsoft SQL Server support

## 1.3.0

* Bump default JIRA version to 6.0.5

## 1.2.0

* Bump default JIRA version to 6.0.2

## 1.1.0

* Bump default JIRA version to 6.0.1

## 1.0.0

* Split default recipe into individual recipes
* apache2 recipe does not include default recipe
* Load database/tomcat settings via Jira.settings library (bonus: help support Chef Solo)
* Moved apache2 attributes into default attributes
* Bump default JIRA version to 5.2.11
* Added url_base attribute
* Auto-detect checksum attribute for some versions
* Added Vagrantfile and Test Kitchen for testing
* minitest fixes
* Added COMPATIBILITY.md
* Refactored README documentation

## 0.1.3

* Chef 11 fixes in apache2 recipe

## v0.1.2

* Hopefully removed hard dependency on java_ark from java cookbook

## v0.1.1

* Added permgen.sh template for custom JAVA_HOME, otherwise always defaults to
  JIRA installed JRE

## v0.1.0

* Initial release


libraries/jira.rb
# Jira module
module Jira
  # Jira::Helpers module
  # rubocop:disable Metrics/ModuleLength
  module Helpers
    # TODO: fix AbcSize
    # rubocop:disable Metrics/AbcSize
    class Jira
      def self.settings(node)
        begin
          if Chef::Config[:solo]
            begin
              settings = Chef::DataBagItem.load('jira', 'jira')['local']
            rescue
              Chef::Log.info('No jira data bag found')
            end
          else
            begin
              settings = Chef::EncryptedDataBagItem.load('jira', 'jira')[node.chef_environment]
            rescue
              Chef::Log.info('No jira encrypted data bag found')
            end
          end
        ensure
          settings ||= node['jira'].to_hash

          case settings['database']['type']
          when 'mysql'
            settings['database']['port'] ||= 3306
          when 'postgresql'
            settings['database']['port'] ||= 5432
          else
            warn 'Unsupported database type! - Use a supported type or handle DB creation/config in a wrapper cookbook!'
          end
        end

        settings
      end
    end
    # rubocop:enable Metrics/AbcSize

    def jira_virtual_host_name
      node['jira']['apache2']['virtual_host_name'] || node['fqdn']
    end

    def jira_virtual_host_alias
      node['jira']['apache2']['virtual_host_alias'] || node['hostname']
    end

    # Detects the current JIRA version.
    # Returns nil if JIRA isn't installed.
    #
    # @return [String] JIRA version
    def jira_version
      pom_file = File.join(
        node['jira']['install_path'],
        '/atlassian-jira/META-INF/maven/com.atlassian.jira/atlassian-jira-webapp/pom.properties'
      )

      begin
        return Regexp.last_match(1) if File.read(pom_file) =~ /^version=(.*)$/
      rescue Errno::ENOENT
        # JIRA is not installed
        return nil
      end
    end

    # Returns download URL for JIRA artifact
    # rubocop:disable Metrics/AbcSize
    # rubocop:disable CyclomaticComplexity
    def jira_artifact_url
      return node['jira']['url'] unless node['jira']['url'].nil?

      base_url = 'https://www.atlassian.com/software/jira/downloads/binary'
      version  = node['jira']['version']

      # JIRA versions >= 7.0.0 have different flavors
      # Also (at this time) the URLs for flavors unfortunately differ
      if Gem::Version.new(version) < Gem::Version.new(7)
        product = "#{base_url}/atlassian-jira-#{version}"
      else
        case node['jira']['flavor']
        when 'software'
          product = "#{base_url}/atlassian-jira-#{node['jira']['flavor']}-#{version}-jira-#{version}"
        when 'core'
          product = "#{base_url}/atlassian-jira-#{node['jira']['flavor']}-#{version}"
        end
      end

      # Return actual URL
      case node['jira']['install_type']
      when 'installer'
        "#{product}-#{jira_arch}.bin"
      when 'standalone'
        "#{product}.tar.gz"
      when 'war'
        fail 'WAR install type is no longer supported by Atlassian and removed from this cookbook.'
      end
    end
    # rubocop:enable CyclomaticComplexity
    # rubocop:enable Metrics/AbcSize

    # Returns SHA256 checksum of specific JIRA artifact
    # rubocop:disable Metrics/AbcSize
    def jira_artifact_checksum
      return node['jira']['checksum'] unless node['jira']['checksum'].nil?

      version = node['jira']['version']
      flavor  = node['jira']['flavor']

      if Gem::Version.new(version) < Gem::Version.new(7)
        sums = jira_checksum_map[version]
      else
        versionsums = jira_checksum_map[version]
        sums = versionsums[flavor]
      end

      fail "JIRA version #{version} is not supported by the cookbook" unless sums

      case node['jira']['install_type']
      when 'installer' then sums[jira_arch]
      when 'standalone' then sums['tar']
      end
    end
    # rubocop:enable Metrics/AbcSize

    def jira_arch
      (node['kernel']['machine'] == 'x86_64') ? 'x64' : 'x32'
    end

    # rubocop:disable Metrics/MethodLength
    # Returns SHA256 checksum map for JIRA artifacts
    def jira_checksum_map
      {
        '5.2' => {
          'x32' => '05335145554cc446adfb90e318139f9408b41cacb57d90d0e11b788718ed8734',
          'x64' => '95841d1b222db63c5653b81583837a5d90e315a2ff2661534b310113afcff33f',
          'tar' => '80e05e65778ce2e3d91422da9d30a547e174fe2b9ba920b7dcdff78e29f353da'
        },
        '5.2.11' => {
          'x32' => '7088a7d123e263c96ff731d61512c62aef4702fe92ad91432dc060bab5097cb7',
          'x64' => 'ad4a851e7dedd6caf3ab587c34155c3ea68f8e6b878b75a3624662422966dff4',
          'tar' => '8d18b1da9487c1502efafacc441ad9a9dc55219a2838a1f02800b8a9a9b3d194'
        },
        '6.0' => {
          'x32' => '3967f8f663c24ff51bf30300ae0f2fb27320c9b356ed4dbf78fce7cc1238eccb',
          'x64' => '915b773c9870ebacbee4712e26d1c9539b48210b6055cb1b2d81e662deae2e60',
          'tar' => '791a8a4a65e40cd00c1ee2a3207935fbcc2c103416e739ad4e3ed29e39372383'
        },
        '6.0.1' => {
          'x32' => 'e383961667e6ef6b5bc387123fa76620a5bdf71413283de5b79cd0ae71248922',
          'x64' => '4e56ef7980b8f3b5b434a7a440d663b9d08e5588d635214e4434eabc3a8d9623',
          'tar' => '492e46119310f378d7944dea0a92c470f0d0b794219d6647a92ea08f8e99f80e'
        },
        '6.0.2' => {
          'x32' => 'bfa7d8731ef2ec5b7e802c119d5a68b1b93505d904c831801236eacff9fa1f5e',
          'x64' => 'fee8fe6804ace532abb805eea5ae0df342526eaf45b2c3e8e34978c97b5aa3aa',
          'tar' => '89b0178bf33488040c032d678ffcdeebc9b9d4565599a31b35515e3aaa391667'
        },
        '6.0.3' => {
          'x32' => 'cda9499247c43c0f812bd2924e569ba3dd08c088e03455ec9c1f79bd30c1509a',
          'x64' => 'cdbd679e70097120c0083e9e0949c66b842742a3a4ccbae0db01b81d9e9fce9e',
          'tar' => '0f94b9d31b8825e91c05e06538dce5891801b83549adbc1dfd26f5b9100c24cf'
        },
        '6.0.4' => {
          'x32' => '4f60c69a13d3d66b0864849d9d3d5a8dfe240830b332cdd8848ae14055709984',
          'x64' => 'd7b845cb21461f032e1563e40f7daa220277809c53e14e4342728f04d0fa039a',
          'tar' => 'ca0f80c36ab408131e283b5c00aead949ce37c4ef8a870b2726eb55882ea6821'
        },
        '6.0.5' => {
          'x32' => '10f9cc9ebb702f01d44f315eabfa4bc1af75dadf161a2cf6d5439c720d604fed',
          'x64' => '0826bf54c7765b053e571d3118b8b48f899d60a76518dc4df34da14a66930e37',
          'tar' => '9050297a28059468a9a3ddfcc8b788aaf62210b341f547d4aebbab92baa96dd3'
        },
        '6.0.6' => {
          'x32' => '030f25f6ab565d66b9f390dced8cafafe0d338ea792d942a6d1901888fa91b7d',
          'x64' => 'bf7145fbbbe0446f3a349e85b7b1277cab3cbe1dfc85029a2fb974f8fac3be59',
          'tar' => '27e699692e107a9790926d5f6fb0ddb89a1bd70e1d6877ce23991c0701495d67'
        },
        '6.0.7' => {
          'x32' => '159f143a1d15c9764b05f0b07f5ea24e4afe851f276ba5290eadabcf5f404a53',
          'x64' => '89da53718d80aad4680e48559ff126ffb35addccfed556c022d5450fb8e44cbb',
          'tar' => '6de5ac1a06116de2c95d5944eab1da416170e8b6bea3a0a7a641b52836100946'
        },
        '6.0.8' => {
          'x32' => 'ad1d17007314cf43d123c2c9c835e03c25cd8809491a466ff3425d1922d44dc0',
          'x64' => 'b7d14d74247272056316ae89d5496057b4192fb3c2b78d3aab091b7ba59ca7aa',
          'tar' => '2ca0eb656a348c43b7b9e84f7029a7e0eed27eea9001f34b89bbda492a101cb6'
        },
        '6.1' => {
          'x32' => 'c879e0c4ba5f508b4df0deb7e8f9baf3b39db5d7373eac3b20076c6f6ead6e84',
          'x64' => '72e49cc770cc2a1078dd60ad11329508d6815582424d16836efd873f3957e2c8',
          'tar' => 'e63821f059915074ff866993eb5c2f452d24a0a2d3cf0dccea60810c8b3063a0'
        },
        '6.1.5' => {
          'x32' => 'f3e589fa34182195902dcb724d82776005a975df55406b4bd5864613ca325d97',
          'x64' => 'b0b67b77c6c1d96f4225ab3c22f31496f356491538db1ee143eca0a58de07f78',
          'tar' => '6e72f3820b279ec539e5c12ebabed13bb239f49ba38bb2e70a40d36cb2a7d68f'
        },
        '6.3.15' => {
          'x32' => '739ac3864951b06a4ce910826f5175523b4ab9eae5005770cbcb774cc94e2e29',
          'x64' => 'a334865dd0b5df5b3bcc506b5c40ab7b65700e310edb6e7e6f86d30c3a8e3375',
          'tar' => '056553ec88cdeeefec73a6692d270a21b9b395af63a5c1ad9865752928dcec2c'
        },
        '6.4.6' => {
          'x32' => 'bede3c18bced84a4b2134ad07c5c4387f6c6991cfaf59768307a31bf72ba8de4',
          'x64' => '0ea1cc37b7de135315b2b241992fca572f808337b730ad68dc0c8c514136a480',
          'tar' => '9bfdba6975cc5188053efe07787d290c12347b62ae13a10d37dd44f14fe68e05'
        },
        '6.4.7' => {
          'x32' => '8545173ce7c0abdad2213a9514adc2b91443acbed31de1a47a385e52521f7114',
          'x64' => '95db7901de1f0c3d346b6ce716cbdf8cd7dc8333024c26b4620be78ba70f3212',
          'tar' => 'c8623ca2a1c0fea18e3921ee1834b3ffe39d70ee2c539f99a99eee2cfb09edd4'
        },
        '6.4.11' => {
          'x32' => 'c68ac38ff0495084dd74d73a85c5e37889af265f3097149a05e4752279610ad6',
          'x64' => '4030010efd5fbec3735dc3a585cd833af957cf7efe4f4bbc34b17175ff9ba328',
          'tar' => 'a8fb59ea41a65e751888491e4c8c26f8a0a6df053805a1308e2b6711980881ec'
        },
        '6.4.12' => {
          'x32' => 'dc807ebed5065416eebb117c061aa57bd07c1d168136aca786ae2b0c100f7e30',
          'x64' => '9897ae190a87a61624d5a307c428e8f4c86ac9ff03e1a89dbfb2da5f6d3b0dbd',
          'tar' => 'a77cf4c646d3f49d3823a5739daea0827adad1254dae1d1677c629e512a7afd4'
        },
        '7.0.0' => {
          'core' => {
            'x32' => 'bcd4746dcd574532061f79ec549e16d8641346f4e45f1cd3db032730fd23ea80',
            'x64' => '314bb496b7d20fb1101eb303c48a80041775e4fadd692fd97583b9c248df5099',
            'tar' => '56bdae7b78ac4472e6c9a22053e4b083d9feb07ee948f4e38c795591d9fc9ae9'
          },
          'software' => {
            'x32' => '3a43274bc2ae404ea8d8c2b50dcb00cc843d03140c5eb11de558b3025202a791',
            'x64' => '49e12b2ba9f1eaa4ed18e0a00277ea7be19ffd6c55d4a692da3e848310815421',
            'tar' => '2eb0aff3e71272dc0fd3d9d6894f219f92033d004e46b25b542241151a732817'
          }
        },
        '7.0.2' => {
          'core' => {
            'x32' => '483cbe3738c5b556ddbadf11adaf98428b0d6d7aec2460eba639c8f4190a6df6',
            'x64' => 'cda659e4b15eb6c70b2ad81acb2917ab66f6a6b114e8f3dad69683ec21b3a844',
            'tar' => '5568de1e67cbfe6c1d3e28869988c78fdc632c59774908d4e229aab1439d255f'
          },
          'software' => {
            'x32' => '235cd2466e3b1e3ac2f4826ee37d64cf53af3c49d72a816a380979931b9fb5fd',
            'x64' => '8ebd0609b3520dfa399672dd10556cbe4886aeb8c59dbf11058b61d5eedb5e2f',
            'tar' => '49a4aca54a5461762d5064b27fce9cb2b8a8a020c1c073c7499a48c19cc8542b'
          }
        }
      }
    end
    # rubocop:enable Metrics/MethodLength
  end
  # rubocop:enable Metrics/ModuleLength
end

::Chef::Recipe.send(:include, Jira::Helpers)
::Chef::Resource.send(:include, Jira::Helpers)
::Chef::Mixin::Template::TemplateContext.send(:include, Jira::Helpers)



recipes/database.rb
settings = Jira.settings(node)

database_connection = {
  :host => settings['database']['host'],
  :port => settings['database']['port']
}

case settings['database']['type']
when 'mysql'

  # We used to include these recipes here - these have since become library cookbooks, and
  # so these includes are unnecessary - and will in fact cause failures / are incorrect.
  # include_recipe 'mysql::server'
  # include_recipe 'database::mysql'

  mysql_client 'default' do
    action :create
  end

  mysql2_chef_gem 'default' do
    action :install
  end

  database_connection.merge!(:username => 'root', :password => node['mysql']['server_root_password'])

  mysql_database settings['database']['name'] do
    connection database_connection
    collation 'utf8_bin'
    encoding 'utf8'
    action :create
  end

  # See this MySQL bug: http://bugs.mysql.com/bug.php?id=31061
  mysql_database_user '' do
    connection database_connection
    host 'localhost'
    action :drop
  end

  mysql_database_user settings['database']['user'] do
    connection database_connection
    host '%'
    password settings['database']['password']
    database_name settings['database']['name']
    action [:create, :grant]
  end
when 'postgresql'
  include_recipe 'postgresql::server'
  include_recipe 'database::postgresql'
  database_connection.merge!(:username => 'postgres', :password => node['postgresql']['password']['postgres'])

  postgresql_database settings['database']['name'] do
    connection database_connection
    connection_limit '-1'
    encoding 'utf8'
    action :create
  end

  postgresql_database_user settings['database']['user'] do
    connection database_connection
    password settings['database']['password']
    database_name settings['database']['name']
    action [:create, :grant]
  end
end


recipes/container_server_configuration.rb
settings = Jira.settings(node)

case node['jira']['container_server']['name']
when 'tomcat'
  if node['jira']['install_type'] == 'war'
    # Only should configure context pointing to war file

    template "#{node['jira']['context_path']}/#{node['jira']['context']}.xml" do
      source 'tomcat/context.xml.erb'
      notifies :restart, 'service[tomcat]', :delayed
      only_if "test -f #{node['jira']['war']['file']}"
    end
  else
    # Installer and standalone installs bundle a Tomcat instance
    # This section customizes said Tomcat instance

    template "#{node['jira']['install_path']}/bin/permgen.sh" do
      source 'tomcat/permgen.sh.erb'
      owner node['jira']['user']
      mode '0755'
      notifies :restart, 'service[jira]', :delayed
    end

    template "#{node['jira']['install_path']}/bin/setenv.sh" do
      source 'tomcat/setenv.sh.erb'
      owner node['jira']['user']
      mode '0755'
      notifies :restart, 'service[jira]', :delayed
    end

    jira_major_version = Gem::Version.new(node['jira']['version']).segments.first
    template_partial_version = (jira_major_version == 7) ? '7' : 'default'

    template "#{node['jira']['install_path']}/conf/server.xml" do
      source 'tomcat/server.xml.erb'
      owner node['jira']['user']
      mode '0640'
      variables({
        tomcat: settings['tomcat'],
        template_partial_version: template_partial_version
      })
      notifies :restart, 'service[jira]', :delayed
    end
  end
end


attributes/default.rb
default['jira']['home_path']          = '/var/atlassian/application-data/jira'
default['jira']['install_path']       = '/opt/atlassian/jira'
default['jira']['install_type']       = 'installer'
default['jira']['version']            = '7.0.2'
default['jira']['flavor']             = 'software'
default['jira']['user']               = 'jira'
default['jira']['group']              = 'jira'
default['jira']['backup_when_update'] = false
default['jira']['ssl']                = false

# Defaults are automatically selected from version via helper functions
default['jira']['url']      = nil
default['jira']['checksum'] = nil

default['jira']['apache2']['access_log']         = ''
default['jira']['apache2']['error_log']          = ''
default['jira']['apache2']['port']               = 80

# Defaults are automatically selected from fqdn and hostname via helper functions
default['jira']['apache2']['virtual_host_name']  = nil # node['fqdn']
default['jira']['apache2']['virtual_host_alias'] = nil # node['hostname']


default['jira']['apache2']['ssl']['access_log']       = ''
default['jira']['apache2']['ssl']['error_log']        = ''
default['jira']['apache2']['ssl']['chain_file']       = ''
default['jira']['apache2']['ssl']['port']             = 443

case node['platform_family']
when 'rhel'
  default['jira']['apache2']['ssl']['certificate_file'] = '/etc/pki/tls/certs/localhost.crt'
  default['jira']['apache2']['ssl']['key_file']         = '/etc/pki/tls/private/localhost.key'
else
  default['jira']['apache2']['ssl']['certificate_file'] = '/etc/ssl/certs/ssl-cert-snakeoil.pem'
  default['jira']['apache2']['ssl']['key_file']         = '/etc/ssl/private/ssl-cert-snakeoil.key'
end

default['jira']['database']['host']     = '127.0.0.1'
default['jira']['database']['name']     = 'jira'
default['jira']['database']['password'] = 'changeit'
default['jira']['database']['type']     = 'mysql'
default['jira']['database']['user']     = 'jira'

# Needed for postgresql unfortunately
if node['jira']['database']['type'] == 'postgresql'
  case node['platform_family']
  when 'debian'
    default['apt']['compile_time_update'] = true
  end
end

# Default is automatically selected from database type via helper function
default['jira']['database']['port'] = nil

default['jira']['jvm']['minimum_memory']  = '256m'
default['jira']['jvm']['maximum_memory']  = '768m'
default['jira']['jvm']['maximum_permgen'] = '256m'
default['jira']['jvm']['java_opts']       = ''
default['jira']['jvm']['support_args']    = ''

default['jira']['tomcat']['port'] = '8080'

default['jira']['crowd_sso']['enabled']        = false
default['jira']['crowd_sso']['sso_appname']    = 'jira'
default['jira']['crowd_sso']['sso_password']   = 'changethistosomethingsensible'
default['jira']['crowd_sso']['crowd_base_url'] = 'http://localhost:8095/crowd/'
